### PR TITLE
Keep ExecInitJunkFilter() the same as upstream.

### PR DESCRIPTION
**Pull request by pivotal-ning-yu**
Signed-off-by: Gang Xiong <gxiong@pivotal.io>

### CODE REVIEW

**Comment by pivotal-ning-yu** [comment]
@hlinnaka @foyzur please help to review this PR, thanks in advance.

File: src/backend/executor/nodeDML.c [hunk 1]
```diff
@@ -168,11 +168,24 @@ ExecInitDML(DML *node, EState *estate, int eflags)
...
+	 * junkfilter to use the same tuple descriptor as the result relation.
+	 * Otherwise the mismatch of tuple descriptor will cause a break in
+	 * ExecInsert()->reconstructMatchingTupleSlot(). 
+	 */
+	TupleDesc cleanTupType = CreateTupleDescCopy(dmlstate->ps.state->es_result_relation_info->ri_RelationDesc->rd_att);
+
+	ExecSetSlotDescriptor(dmlstate->junkfilter->jf_resultSlot, cleanTupType);
+
+	ReleaseTupleDesc(dmlstate->junkfilter->jf_cleanTupType);
+	dmlstate->junkfilter->jf_cleanTupType = cleanTupType;
```
Comment (foyzur) [inline src/backend/executor/nodeDML.c]: I don't like that the junkfilter->jf_cleanTupType and junkfilter->jf_cleanMap are computed from two different sources; one from targetlist and the other from rd_att. @vraghavan78 @d can the target list of dml be out of sync (mismatched order) from the underlying relation attribute order? I cannot think of an example. The ExecFilterJunk will use cleanMap to map "clean" columns on to the clean slot. So, they are logically tied together.

**Comment by foyzur** [comment]
Except my concern about cleanMap and tuple descriptor computation from two different sources, this PR looks good.

**Comment by d** [comment]
👍 
Mostly looks good to me, but I don't feel entirely confident reviewing changes in those areas, deferring to @foyzur

**Comment by xiong-gang** [comment]
Hi @d @foyzur ,
Can we merge this patch? Or are you planning to handle this differently?
Thanks.

**Comment by d** [comment]
Sorry Gang, we inadvertently dropped the ball here. Please go ahead and integrate this. We should re-open the conversation to further converge this code when ORCA adds type modifier support.

**Comment by pivotal-ning-yu** [comment]
@d thanks for the updates, does this mean that this PR is ready to merge?

**Comment by pivotal-ning-yu** [comment]
rebased to resolve conflicts.

**Comment by pivotal-ning-yu** [comment]
merged.

**Pull request merged**